### ISSUE TITLE
should one be able to return a reference to the result of a downcast?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#14834]
Is the following program legal?

``` chapel


class Parent {
} 

class Child : Parent {
}

var glob: owned Parent = new Child();

proc getit() ref : owned Child? {
  return glob:owned Child?;
}

proc main() {
  var x = getit();
  writeln(x);
  writeln(glob);
}

```

Today, the compiler gives an error, because `glob: owned Child?` creates a new value storing the owned (and in fact transfers it out of `glob`. But the intent here was to get a reference to glob that had a different type. The programmer might know that `glob` has the other type for some reason (e.g. just checked the downcast). Should the programmer be able to return a reference to the original `glob` with the different type? I.e. without doing ownership transfer?

This came up while I was looking at #13600 for TaskErrors.filter.

**Comment by bradcray** [comment]
[This example](https://tio.run/##XY1BDsIgEEX3nGLSFSSNB6AxLly58wrYTikJQkNHW2M8OxaIsbqb/P/@m3ZQI9oYW6umCc4qoCN4shcwVqLjYGwHclMxdlcBtPUXCX522H26PTicy4CLhrEx@BY0kiEuIGAPHz4jh9UFa0y34IptWzbpTxb03vM0Xn7WIq/nYAit49XJJa4G7QmqGpYaKpmPHT1GlBMF47T4Sq/KOF4cyf//PpExvgE) suggests that the answer should potentially be "yes" (?).

But I assume we're only talking about doing this for classes where the downcast refers to the original object still and not cases like integer downcasts where a new value is created.

**Comment by mppf** [comment]
> But I assume we're only talking about doing this for classes where the downcast refers to the original object still and not cases like integer downcasts where a new value is created.

Right. In any case we currently have some issues in this area (#8489 being what I'd consider the most obvious) and if we allow it it would only be for classes.

(For now, I plan to apply a workaround to `Error`).

**Comment by cassella** [comment]
I'm confused by the example having `getit()` return a nilable type.  If the behavior is changed as described, would that let `main()` set `x = nil;`?  Would that result in the non-nilable `glob` being set to `nil`?

**Comment by mppf** [comment]
@cassella - that's an excellent point and the nilability part doesn't need to be part of this discussion and so I've updated the example.

I'll elaborate a bit on what you were saying. One shouldn't be able to return a reference to the result of an upcast, to nilable or otherwise:
``` chapel
var c: borrowed Child = new Child();
proc f() ref : borrowed Parent {
  return c : borrowed Parent;
}
```

Even though we do not need to create new storage to perform the cast (or coercion), the caller could use this code to change `c` to store something of type `borrowed Parent`.

But, `const ref` does not have this problem.

**Comment by mppf** [comment]
Revisiting the original example program:

``` chapel
class Parent { } 

class Child : Parent {  var y: int; }

var glob: owned Parent = new Child();

proc getit() ref : owned Child {
  return glob:owned Child;
}

ref refGlobal = getit();
```

Now, imagine that `glob` is updated while this reference exists:

``` chapel
glob = new Parent();
writeln(refGlobal.y); // uh... but now refGlobal points to a Parent that doesn't have a y ?
```

As a result this functionality would invalidate the type system.
